### PR TITLE
musl libc support

### PR DESCRIPTION
**Pull request by SamuelMarks**
Warnings are bad! - Found one in your library deep within another:
```
#17 120.2   In file included from /tmp/wd/postgresml/pgml-extension/target/debug/build/xgboost-sys-a1e545b7b843ddb4/out/xgboost/rabit/include/rabit/internal/socket.h:38,
#17 120.2                    from /tmp/wd/postgresml/pgml-extension/target/debug/build/xgboost-sys-a1e545b7b843ddb4/out/xgboost/src/collective/loop.cc:15:
#17 120.2   /usr/include/sys/poll.h:1:2: warning: #warning redirecting incorrect #include <sys/poll.h> to <poll.h> [-Wcpp]
#17 120.2       1 | #warning redirecting incorrect #include <sys/poll.h> to <poll.h>
#17 120.2         |  ^~~~~~~
#17 120.2   In file included from /tmp/wd/postgresml/pgml-extension/target/debug/build/xgboost-sys-a1e545b7b843ddb4/out/xgboost/rabit/include/rabit/internal/socket.h:38,
#17 120.2                    from /tmp/wd/postgresml/pgml-extension/target/debug/build/xgboost-sys-a1e545b7b843ddb4/out/xgboost/src/collective/socket.cc:14:
#17 120.2   /usr/include/sys/poll.h:1:2: warning: #warning redirecting incorrect #include <sys/poll.h> to <poll.h> [-Wcpp]
#17 120.2       1 | #warning redirecting incorrect #include <sys/poll.h> to <poll.h>
#17 120.2         |  ^~~~~~~
#17 120.2   In file included from /tmp/wd/postgresml/pgml-extension/target/debug/build/xgboost-sys-a1e545b7b843ddb4/out/xgboost/rabit/include/rabit/internal/socket.h:38,
#17 120.2                    from /tmp/wd/postgresml/pgml-extension/target/debug/build/xgboost-sys-a1e545b7b843ddb4/out/xgboost/src/collective/tracker.cc:4:
#17 120.2   /usr/include/sys/poll.h:1:2: warning: #warning redirecting incorrect #include <sys/poll.h> to <poll.h> [-Wcpp]
#17 120.2       1 | #warning redirecting incorrect #include <sys/poll.h> to <poll.h>
#17 120.2         |  ^~~~~~~
```

EDIT0: Another one, resolved it same way they did https://github.com/gperftools/gperftools/issues/693' ; a simple `#define mmap64 mmap`

EDIT1: To replicate, here's a `Dockerfile`:
```dockerfile
FROM alpine
RUN apk add --no-cache cmake gcc g++ make musl-dev
WORKDIR /src/xgboost
COPY . .
RUN cmake -S . -B build && \
    cmake --build build
```

### CODE REVIEW

File: include/xgboost/unixdefs.h [hunk 1]
```diff
@@ -0,0 +1,14 @@
+#if defined(UNIXDEFS_H) && !defined(_MSC_VER) && \
+    !defined(__MINGW32__) && !defined(__CYGWIN__) && \
+    !defined(_WIN32) && !defined(__WINDOWS__) && \
+    !defined(__MUSL__)
+#define UNIXDEFS_H
+
+#define _GNU_SOURCE
+#include <features.h>
```
Comment (trivialfis) [inline include/xgboost/unixdefs.h:8]: Could you please share why we need this header?

File: include/xgboost/unixdefs.h [hunk 2]
```diff
@@ -0,0 +1,14 @@
+#if defined(UNIXDEFS_H) && !defined(_MSC_VER) && \
+    !defined(__MINGW32__) && !defined(__CYGWIN__) && \
+    !defined(_WIN32) && !defined(__WINDOWS__) && \
+    !defined(__MUSL__)
+#define UNIXDEFS_H
+
+#define _GNU_SOURCE
```
[the remark this answers is not in the document]
Comment (SamuelMarks) [reply include/xgboost/unixdefs.h:7]: https://stackoverflow.com/a/5583764
> Defining `_GNU_SOURCE` has nothing to do with license and everything to do with writing (non-)portable code.

File: include/xgboost/unixdefs.h (hunk 1 above)
Comment (SamuelMarks) [reply include/xgboost/unixdefs.h:8]: Feature flags. Can put it in another header if you think it works better semantically there.

File: include/xgboost/collective/poll_utils.h [hunk 3]
```diff
@@ -35,7 +35,12 @@
 
 #if !defined(_WIN32)
 
+#include "xgboost/unixdefs.h"
+#ifdef __MUSL__
+#include <poll.h>
+#else
 #include <sys/poll.h>
```
Comment (trivialfis) [inline include/xgboost/collective/poll_utils.h:42]: Apologies for the wrong import. I think `#include <poll.h>` should work fine given https://man7.org/linux/man-pages/man2/poll.2.html

File: src/common/io.cc [hunk 4]
```diff
@@ -7,6 +7,11 @@
 #include <sys/mman.h>  // for mmap, mmap64, munmap, madvise
 #include <unistd.h>    // for close, getpagesize
 
+#include "xgboost/unixdefs.h"
+#ifdef __MUSL__
+#define mmap64 mmap
```
Comment (trivialfis) [inline src/common/io.cc:12]: We can use `mmap` as default and let the 32-bit system raise an overflow error when the file size is larger than 2GB.

**Review by trivialfis — COMMENTED** [review]
Thank you for the replies. Looking closer, I think the musl folks are right and there's no need to make specialization for musl. Please see the inlined comments for the proposed changes.

**Comment by SamuelMarks** [comment]
> Thank you for the replies. Looking closer, I think the musl folks are right and there's no need to make specialization for musl. Please see the inlined comments for the proposed changes.

@trivialfis Yes that is correct; if you use `poll.h` and `mmap` then you won't need this PR. As for the `mmap` fallout, you can also check the other direction, if `mmap64` is available then `_LARGEFILE64_SOURCE` is defined or more correctly `#if defined(_FILE_OFFSET_BITS) && _FILE_OFFSET_BITS >= 64` then you can use that symbol. See also: https://pubs.opengroup.org/onlinepubs/009695399/utilities/c99.html#tag_04_12_13_03

Otherwise sure you can just switch to `mmap` over `mmap64` for everyone. It's your call*.
*pun intended?

**Comment by trivialfis** [comment]
> Yes that is correct; if you use poll.h and mmap then you won't need this PR

Sounds great! That should make things much simpler. Would you like to help make the fix?

> As for the mmap fallout, you can also check the other direction

I don't think that's necessary, users of 32-bit platforms can reduce the batch size to be smaller than 2 GB.

**Comment by SamuelMarks** [comment]
@trivialfis Sure thing done in #10767

**Pull request closed without merging**